### PR TITLE
RUMM-2464 [SR] Support incremental updates

### PR DESCRIPTION
**Pull request by ncreated**
### What and why?

📦 This PR adds support to incremental updates in SR segments.

This will vastly reduce the size of SR payloads. Instead of sending all wireframes each time ("full snapshot"), now we only send their differences from previous frame ("incremental snapshot").

### How?

This work is composed of two main parts:
* wireframe IDs generation,
* diff algorithm.

#### 📷 Wireframe IDs generation

I introduced `NodeIDGenerator` which generates unique and stable IDs (`NodeID`) for each `UIView` instance:
* **`NodeIDs` are unique**, so each instance of `UIView` has different ID(s);
* **`NodeIDs` are stable**, meaning that one instance of `UIView` has the same ID for its entire lifespan.

`NodeIDs` are generated by `Recorded` (UI thread) and used as `WireframeID` in `Processor` (background thread). They are keys to determine incremental changes from one array of wireframes to another.

#### 🔃 Diff algorithm

I added generic diff function:
```swift
internal protocol Diffable {
    var id: DiffableID { get }
    func isDifferent(than otherElement: Self) -> Bool
}

internal func computeDiff<E: Diffable>(oldArray: [E], newArray: [E]) throws -> Diff<E> {
```
which is further specialised by `SRWireframe: Diffable` conformance. Making it generic vastly helped for code isolation and testability (I managed to cover most cases + added property-based and fuzzy tests on complex sequences).

The algorithm itself is based on [Paul Heckel's paper: _"A technique for isolating differences between files" (1978)_](https://dl.acm.org/citation.cfm?id=359467). It is fast and performant (`O(n)` in time and memory) way of computing changes in text files. It's quite popular and there are many implementations out there (e.g. [DifferenceKit](https://github.com/ra1028/DifferenceKit), or `hdiff` in Rust). I managed to simplify it a bit for our problem - unlike lines in text files, `IDs` in our wireframes are guaranteed to be unique.

### Review checklist
- [ ] Feature or bugfix MUST have appropriate tests (unit, integration)
- [ ] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [ ] Run unit tests
- [ ] Run integration tests
- [ ] Run smoke tests

### CODE REVIEW

File: session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+/// - _"A technique for isolating differences between files"_ Paul Heckel (1978) - https://dl.acm.org/citation.cfm?id=359467
+///
+/// - Parameters:
+///   - oldArray: original array
+///   - newArray: new array
+/// - Returns: `Diff` describing changes from `oldArray` to `newArray`.
+internal func computeDiff<E: Diffable>(oldArray: [E], newArray: [E]) throws -> Diff<E> {
+    var table: [DiffableID: Symbol] = [:]
+    var oa: [Entry] = [] // old array entries
+    var na: [Entry] = [] // new array entries
```
Comment (ncreated) [inline session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift:118]: Both `table`, `oa` and `na` names refer to the Heckel's paper and are commonly used among different implementations that I found around. I'd keep it for clarity even if it's not our convention.

File: session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+/// - when reconstructing `newArray` from `oldArray` and `Diff` **removals must be applied before additions**;
+/// - if certain element changes its position in `newArray`, `Diff` includes its **removal and addition** (on new position).
+internal struct Diff<Element: Diffable> {
+    /// Dictates that new element was added in `newArray`.
+    struct Add {
+        /// The `id` of preceding element in `newArray` or `nil` if new element should be inserted at index `0`.
+        let previousID: DiffableID?
+        /// The element to add.
+        let new: Element
+    }
```
Comment (ncreated) [inline session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift:41]: As discussed, it's gonna be `nextID: DiffableID?` in place of `previousID`. It's not a big change (and no change to the algorithm itself). It can be updated in another PR.

File: session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift (hunk 1 above)
Comment (maxep) [reply session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift:118]: For algorithm like this, short var name helps for readability 👍

File: session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+/// Note: Unlike original algorithm, our implementation doesn't use `OC`, `NC` and `OLNO` counters. In our case, elements within each array are unique,
+/// so instead of repetition counters, we define basic `Bool` flags to track occurence of certain `id` in one or both files.
+private struct Symbol {
+    /// If element with certain `id` occurs in `oldArray`.
+    var inOld: Bool
+    /// If element with certain `id` occurs in `newArray`.
+    var inNew: Bool
+    /// The index of element in `oldArray`.
+    var indexInOld: Int?
+}
```
Comment (maxep) [inline session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift:85]: If `indexInOld` is nil it means that symbol is not in old. Then, this could be:
```swift
private struct Symbol {
    /// If element with certain `id` occurs in `newArray`.
    var inNew: Bool
    /// The index of element in `oldArray`.
    var indexInOld: Int?
}
```

It would simplify the condition below in `computeDiff`:
```swift
if symbol.inOld && symbol.inNew {
    guard let indexInOld = symbol.indexInOld else {
        throw DiffError()
    }
    na[index] = .index(indexInOld)
    oa[indexInOld] = .index(index)
}
```
to 
```swift
if symbol.inNew, let indexInOld = symbol.indexInOld {
    na[index] = .index(indexInOld)
    oa[indexInOld] = .index(index)
}
```

File: session-replay/Sources/DatadogSessionReplay/Processor/SRDataModelsBuilder/RecordsBuilder.swift [hunk 4]
```diff
@@ -36,14 +36,62 @@ internal class RecordsBuilder {
...
+        if let lastWireframes = lastWireframes {
+            do {
+                return try createIncrementalRecord(from: snapshot, newWireframes: wireframes, lastWireframes: lastWireframes)
+            } catch { // TODO: RUMM-2410 Use `DD.logger` and / or `DD.telemetry` to report ISR errors
+                // In case of any trouble, fallback to FSR which is always possible:
+                return createFullSnapshotRecord(from: snapshot, wireframes: wireframes)
+            }
+        } else {
+            return createFullSnapshotRecord(from: snapshot, wireframes: wireframes)
+        }
```
Comment (maxep) [inline session-replay/Sources/DatadogSessionReplay/Processor/SRDataModelsBuilder/RecordsBuilder.swift:56]: `guard` would be best here :)

File: session-replay/Sources/DatadogSessionReplay/Processor/SRDataModelsBuilder/RecordsBuilder.swift [hunk 5]
```diff
@@ -36,14 +36,62 @@ internal class RecordsBuilder {
...
+
+        let record = SRIncrementalSnapshotRecord(
+            data: .mutationData(
+                value: .init(
+                    adds: diff.adds.map { addition in .init(previousId: addition.previousID, wireframe: addition.new) },
+                    removes: diff.removes.map { removal in .init(id: removal.id) },
+                    updates: try diff.updates.compactMap { update in
+                        let newWireframe = update.to
+                        let oldWireframe = update.from
+                        return try newWireframe.mutations(from: oldWireframe)
```
Comment (maxep) [inline session-replay/Sources/DatadogSessionReplay/Processor/SRDataModelsBuilder/RecordsBuilder.swift:86]: lets save copying elements
```suggestion
                        try update.to.mutations(from: update.from)
```

File: session-replay/Sources/DatadogSessionReplay/Recorder/SnapshotBuilder/NodeIDGenerator.swift [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            let id = getNextID()
+            view.nodeID = id
+            return id
+        }
+    }
+
+    /// Returns two `NodeIDs` for given instance of `UIView`.
+    /// - Parameter view: the `UIView` object
+    /// - Returns: `NodeIDs` of queried instance
+    func nodeID2(for view: UIView) -> NodeID2 {
```
Comment (maxep) [inline session-replay/Sources/DatadogSessionReplay/Recorder/SnapshotBuilder/NodeIDGenerator.swift:52]: IMO, It would be clearer if the method returns tuples explicitly.
`func nodeID2(for view: UIView) -> (NodeID, NodeID)`

Comment (maxep) [reply session-replay/Sources/DatadogSessionReplay/Recorder/SnapshotBuilder/NodeIDGenerator.swift:52]: Also, this won't scale well if we have components that have other requirements.
We could have a method for requesting `n` IDs.
```func nodeIDs(for view: UIView, count: Int) -> [NodeID]```

**Review by maxep — COMMENTED** [review]
Using this algorithm is simply brillant, well done on translating this to our use case.
I've left some suggestion but it looks great!

File: session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+                oa[indexInOld] = .index(index)
+            }
+        }
+    }
+
+    // 4th pass
+    // > If a line has been found to be unaltered, and the lines immediately adjacent to it in both files are identical,
+    // > then these lines must be the same line. This information can be used to find blocks of unchanged lines.
+    if na.count > 1 {
+        for i in (1..<(na.count - 1)) {
```
Comment (mariusc83) [inline session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift:163]: Why are you starting from 1 here ?

File: session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+                    oa[j + 1] = .index(i + 1)
+                }
+            }
+        }
+    }
+
+    // 5th pass
+    // Similar to 4th pass, except it processes entries in descending order.
+    if na.count > 1 {
+        for i in (1..<(na.count - 1)).reversed() {
```
Comment (mariusc83) [inline session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift:176]: here you are missing the last element of na. `na[count-1]`. Is this intended ?

File: session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift (hunk 7 above)
Comment (maxep) [reply session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift:163]: Indeed, it looks like the [4th](https://gist.github.com/ndarville/3166060#fourth-pass) and [5th](https://gist.github.com/ndarville/3166060#fifth-pass) pass of the algorithm should process all values of `NA` and `OA`.

File: session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift (hunk 3 above)
Comment (ncreated) [reply session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift]: Good idea, done 👌!

File: session-replay/Sources/DatadogSessionReplay/Processor/SRDataModelsBuilder/RecordsBuilder.swift (hunk 4 above)
Comment (ncreated) [reply session-replay/Sources/DatadogSessionReplay/Processor/SRDataModelsBuilder/RecordsBuilder.swift]: Done 👍👌!

Although, I prefer using `guard x() else { y() }` to express that _"we **expect** `x`, but if it's not there, **fallback** to `y()`"_. In this case `createFSR()` is not a fallback but a regular control flow (_"if `lastWireframes` is not yet there, produce FSR"_). However, it is just my habit and not a shared convention, so I changed it to `guard` which we use more often for similar things.

File: session-replay/Sources/DatadogSessionReplay/Processor/SRDataModelsBuilder/RecordsBuilder.swift (hunk 5 above)
Comment (ncreated) [reply session-replay/Sources/DatadogSessionReplay/Processor/SRDataModelsBuilder/RecordsBuilder.swift]: It wasn't copied 🖐️ - with Swift's copy-on-write optimisation that will use the same memory (unless it would be `var` and modified later). Anyway, changed it for more clarity 👍.

File: session-replay/Sources/DatadogSessionReplay/Recorder/SnapshotBuilder/NodeIDGenerator.swift (hunk 6 above)
Comment (ncreated) [reply session-replay/Sources/DatadogSessionReplay/Recorder/SnapshotBuilder/NodeIDGenerator.swift:52]: I'd prefer to keep it as it gives me good documentation point:
```swift
/// Single unique ID of a view in view-tree hierarchy.
/// It is used to mark `UIViews` which correspond to single wireframe in the replay.
internal typealias NodeID = Int64

/// Pair of unique IDs of a single view in view-tree hierarchy.
/// It is used to mark `UIViews` which correspond to two wireframes in the replay.
internal typealias NodeID2 = (NodeID, NodeID)
```
About scalability: it is true, but I'm not afraid of it - it's extremely unlikely that we'll need more than a few IDs per `UIView`. Even this tuple is quite experimental for now and used only to represent `UISwitch` as two wireframes (this may change if we decide to render OS-specific elements on the player side by introducing OS-specific wireframes instead of mimicking its look&feel with 2 shape WFs).

How does it sound @maxep ?

File: session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift (hunk 7 above)
Comment (ncreated) [reply session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift]: Yeah, I think it's a leftover after safeguarding boundaries of this iteration. It should start from `0` - it's fixed now 👌.

Interestingly, it wasn't causing any trouble even after I added more fuzzy tests on short and long sequences. For the sake of safety, I pushed enhancement to tests as well 👍.

File: session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift (hunk 8 above)
Comment (ncreated) [reply session-replay/Sources/DatadogSessionReplay/Processor/Diffing/Diff.swift]: Indeed, good catch - it should loop through `(1..<na.count)` range - fixed 👍👌. Similar to https://github.com/DataDog/dd-sdk-ios/pull/1005#discussion_r995935265, I couldn't figure out if it causes any problem even after enhancing fuzzy tests.

File: session-replay/Sources/DatadogSessionReplay/Recorder/SnapshotBuilder/NodeIDGenerator.swift (hunk 6 above)
Comment (maxep) [reply session-replay/Sources/DatadogSessionReplay/Recorder/SnapshotBuilder/NodeIDGenerator.swift:52]: Yes sounds good, we just need to be careful not to introduce too many `typealias`es at the module level (inner to object is fine), it can get difficult to track and be confusing.

**Review by maxep — APPROVED** [review]
Well done 👏
